src/tool/subcommands/api_cmd/api_run_tests.rs

```diff
@@ -0,0 +1,463 @@
+            anyhow::ensure!(removed);
+        }
+
+        Ok(())
+    })
+}
+
+fn eth_new_block_filter() -> RpcTestScenario {
+    RpcTestScenario::basic(move |client| async move {
+        let filter_id = client.call(EthNewBlockFilter::request(())?).await?;
+
+        let filter_result = client
+            .call(EthGetFilterChanges::request((filter_id.clone(),))?)
+            .await?;
+
+        let result = if let EthFilterResult::Hashes(prev_hashes) = filter_result {
+            let verify_hashes = async |hashes: &[EthHash]| {
+                for hash in hashes {
+                    let _block = client
+                        .call(EthGetBlockByHash::request((hash.clone(), false))?)
+                        .await?;
+                }
+                Ok::<(), crate::rpc::ClientError>(())
+            };
+            verify_hashes(&prev_hashes).await?;
+
+            // Wait till the next block arrive
+            sleep(Duration::from_secs(EPOCH_DURATION_SECONDS as u64)).await;
+
+            let filter_result = client
+                .call(EthGetFilterChanges::request((filter_id.clone(),))?)
+                .await?;
+
+            if let EthFilterResult::Hashes(hashes) = filter_result {
+                verify_hashes(&hashes).await?;
+                anyhow::ensure!(prev_hashes != hashes);
+
+                Ok(())
+            } else {
+                Err(anyhow::anyhow!("expecting blocks"))
+            }
+        } else {
+            Err(anyhow::anyhow!("expecting blocks"))
+        };
+
+        let removed = client
+            .call(EthUninstallFilter::request((filter_id,))?)
+            .await?;
+        anyhow::ensure!(removed);
+
+        result
+    })
+}
+
+fn eth_new_pending_transaction_filter(tx: TestTransaction) -> RpcTestScenario {
+    RpcTestScenario::basic(move |client| {
+        let tx = tx.clone();
+        async move {
+            let filter_id = client
+                .call(EthNewPendingTransactionFilter::request(())?)
+                .await?;
+
+            let filter_result = client
+                .call(EthGetFilterChanges::request((filter_id.clone(),))?)
+                .await?;
+
+            let result = if let EthFilterResult::Hashes(prev_hashes) = filter_result {
+                let encoded = cbor4ii::serde::to_vec(
+                    Vec::with_capacity(tx.payload.len()),
+                    &Value::Bytes(tx.payload.clone()),
+                )
+                .context("failed to encode params")?;
+
+                let message = Message {
+                    to: tx.to,
+                    from: tx.from,
+                    method_num: EVMMethod::InvokeContract as u64,
+                    params: encoded.into(),
+                    ..Default::default()
+                };
+
+                let smsg = client
+                    .call(MpoolPushMessage::request((message, None))?)
+                    .await?;
+
+                sleep(Duration::from_secs(2)).await;
```

Comment:
`sleeps` in tests are suspicious. They make them slow and flaky (especially when run on shared CPUs). Can we make it more deterministic?

Reply:
Unfortunately, we still need to use some form of sleep in a busy loop to wait for the message to arrive in the pool and avoid any race conditions. Probably we could make use of the subscription API, but it won't work with regular messages.